src/symbol/collision_index.js

```diff
@@ -84,56 +86,34 @@ class CollisionIndex {
-        // This is a quick and dirty solution for chosing which collision circles to use (since collision circles are
-        // laid out in tile units). Ideally, I think we should generate collision circles on the fly in viewport coordinates
-        // at the time we do collision detection.
-        // See https://github.com/mapbox/mapbox-gl-js/issues/5474
-
-        // incidenceStretch is the ratio of how much y space a label takes up on a tile while drawn perpendicular to the viewport vs
-        //  how much space it would take up if it were drawn flat on the tile
-        // Using law of sines, camera_to_anchor/sin(ground_angle) = camera_to_center/sin(incidence_angle)
-        // Incidence angle 90 -> head on, sin(incidence_angle) = 1, no stretch
-        // Incidence angle 1 -> very oblique, sin(incidence_angle) =~ 0, lots of stretch
-        // ground_angle = u_pitch + PI/2 -> sin(ground_angle) = cos(u_pitch)
-        // incidenceStretch = 1 / sin(incidenceAngle)
-
-        const incidenceStretch = pitchWithMap ? 1 : cameraToAnchorDistance / this.pitchfactor;
-        const lastSegmentTile = tileDistance.lastSegmentViewportDistance * pixelsToTileUnits;
-        return tileDistance.prevTileDistance +
-            lastSegmentTile +
-            (incidenceStretch - 1) * lastSegmentTile * Math.abs(Math.sin(lastSegmentAngle));
-    }
-
-    placeCollisionCircles(collisionCircles: Array<number>,
-                          allowOverlap: boolean,
-                          scale: number,
-                          textPixelRatio: number,
+    placeCollisionCircles(allowOverlap: boolean,
                           symbol: any,
                           lineVertexArray: SymbolLineVertexArray,
                           glyphOffsetArray: GlyphOffsetArray,
                           fontSize: number,
                           posMatrix: mat4,
                           labelPlaneMatrix: mat4,
+                          labelToScreenMatrix?: mat4,
                           showCollisionCircles: boolean,
                           pitchWithMap: boolean,
-                          collisionGroupPredicate?: any): { circles: Array<number>, offscreen: boolean } {
+                          collisionGroupPredicate?: any,
+                          circlePixelDiameter: number,
+                          textPixelPadding: number): { circles: Array<number>, offscreen: boolean, collisionDetected: boolean } {
         const placedCollisionCircles = [];
 
-        const projectedAnchor = this.projectAnchor(posMatrix, symbol.anchorX, symbol.anchorY);
+        const perspectiveRatio = this.projectAnchor(posMatrix, symbol.anchorX, symbol.anchorY).perspectiveRatio;
+        const labelPlaneFontSize = pitchWithMap ? fontSize / perspectiveRatio : fontSize * perspectiveRatio;
+        const labelPlaneFontScale = labelPlaneFontSize / 24;
```

Comment:
Is 24 supposed to be representing ONE_EM defined in src/symbol/one_em.js?

Reply:
Yes it is, thanks for pointing it out :)